warp-tls/Network/Wai/Handler/WarpTLS.hs

```diff
@@ -82,6 +84,12 @@ data TLSSettings = TLSSettings {
   , certMemory :: Maybe S.ByteString
   , chainCertsMemory :: [S.ByteString]
   , keyMemory :: Maybe S.ByteString
+  , certRef :: Maybe (I.IORef S.ByteString)
```

Comment:
It seems to me like the logic here is starting to get difficult to follow, comparing `certMemory` vs `certRef` vs `certFile`. What about introducing sum types for the different possibilities?

Reply:
That makes a lot of sense - done in d9c8861. 

I don't think this is a breaking change, although the behaviour in `loadCredentials` is now different. For the `ByteString` options it previously used `maybe` to fall back to filenames, but I don't think it was actually possible to trigger this behaviour since the fields of `TLSSettings` are not exported and thus `tlsSettingsMemory` and `tlsSettingsChainMemory` were the only ways to set `certMemory` and `keyMemory`.